farming/logics/traits/master_chef/farming.rs

```diff
@@ -0,0 +1,600 @@
+                        )
+                        .checked_div(total_alloc_point.into())
+                        .ok_or(FarmingError::DivByZero1)?
+                        .try_into()
+                        .map_err(|_| FarmingError::CastTou128Error1)?,
+                    )
+                    .ok_or(FarmingError::AddOverflow6)?;
+            } else {
+                arsw_reward = arsw_reward
+                    .checked_add(
+                        casted_mul(
+                            (self._period_max(period)? as u128)
+                                .checked_sub(last_block.into())
+                                .ok_or(FarmingError::SubUnderflow5)?
+                                .checked_mul(pool_info.alloc_point.into())
+                                .ok_or(FarmingError::MulOverflow6)?,
+                            self._arsw_per_block(period)? as u128,
+                        )
+                        .checked_div(total_alloc_point.into())
+                        .ok_or(FarmingError::DivByZero2)?
+                        .try_into()
+                        .map_err(|_| FarmingError::CastTou128Error3)?,
+                    )
+                    .ok_or(FarmingError::AddOverflow7)?;
+
+                last_block = self._period_max(period)?;
+            }
+
+            period += 1;
+        }
+        Ok(casted_mul(arsw_reward, ACC_ARSW_PRECISION)
+            .checked_div(lp_supply.into())
+            .ok_or(FarmingError::DivByZero3)?
+            .try_into()
+            .map_err(|_| FarmingError::CastTou128Error6)?)
+    }
+
+    fn _get_period(&self, block_number: u32) -> Result<u32, FarmingError> {
+        ensure!(
+            block_number >= ARTHSWAP_ORIGIN_BLOCK,
+            FarmingError::BlockNumberLowerThanOriginBlock
+        );
+
+        // BLOCK_PER_PERIOD is never 0
+        return Ok(block_number
+            .checked_sub(ARTHSWAP_ORIGIN_BLOCK)
+            .ok_or(FarmingError::SubUnderflow1)?
+            / BLOCK_PER_PERIOD)
+    }
+
+    fn _period_max(&self, period: u32) -> Result<u32, FarmingError> {
+        Ok(ARTHSWAP_ORIGIN_BLOCK
+            .checked_add(
+                BLOCK_PER_PERIOD
+                    .checked_mul(period.checked_add(1).ok_or(FarmingError::AddOverflow4)?)
+                    .ok_or(FarmingError::MulOverflow1)?,
+            )
+            .ok_or(FarmingError::AddOverflow5)?
+            - 1)
+    }
+
+    fn _arsw_per_block(&self, period: u32) -> Result<Balance, FarmingError> {
+        if period > MAX_PERIOD {
+            return Ok(0)
+        }
+        Ok(casted_mul(
+            FIRST_PERIOD_REWERD_SUPPLY,
+            9u128
+                .checked_pow(period)
+                .ok_or(FarmingError::PowOverflow1)?
+                / 10u128
+                    .checked_pow(period)
+                    .ok_or(FarmingError::PowOverflow2)?,
+        )
+        .try_into()
+        .map_err(|_| FarmingError::CastTou128Error5)?)
```

Comment:
```suggestion
        Ok((casted_mul(
            FIRST_PERIOD_REWERD_SUPPLY,
            9u128
                .checked_pow(period)
                .ok_or(FarmingError::PowOverflow1)?,
        ) / 10u128
            .checked_pow(period)
            .ok_or(FarmingError::PowOverflow2)?)
        .try_into()
        .map_err(|_| FarmingError::CastTou128Error5)?)
```
To preserve precision as much as possible.